### ISSUE TITLE
Bug: Unconsistence behavior between cli and api call with vue-plugin

### ISSUE THREAD

**Issue by ziofat** [typescript-eslint/typescript-eslint#9278]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

This is not related to vue-plugin but I need a non-ts file to reproduce this issue.

When I try to lint a vue file with `eslint test.vue` in command line, it will throw an error with message: `Parsing error: "parserOptions.programs" has been provided for @typescript-eslint/parser.`.

But the same config can be used to lint same file when I lint it by calling `ESLint.lintFiles`.

The possible cause is the `parserOptions.programs` created by parser are different based on the `singleRun` option. The programs created when `singleRun` is `true` has no vue files included.

### Reproduction Repository Link

https://github.com/ziofat/tseslint-vue-v8

### Repro Steps

1) clone the repo
2) `npm install`
3) `npm lint:cli` (which will throw error)
4) `npm lint:api` (which will output lint problems as expected)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.0-alpha.20` |
| `@typescript-eslint/parser`            | `8.0.0-alpha.20` |
| `@typescript-eslint/scope-manager`     | `8.0.0-alpha.20` |
| `@typescript-eslint/typescript-estree` | `8.0.0-alpha.20` |
| `@typescript-eslint/type-utils`        | `8.0.0-alpha.20` |
| `@typescript-eslint/utils`             | `8.0.0-alpha.20` |
| `TypeScript`                           | `5.4.5` |
| `ESLint`                               | `9.3.0` |
| `node`                                 | `18.20.2` |

**Comment by ziofat** [comment]
For now, a working solution is to add `disallowAutomaticSingleRunInference: true` in `parserOptions`.

**Comment by JoshuaKGoldberg** [comment]
👋 sorry for the late response here @ziofat - but I believe this is working as intended.

`singleRun` inference as controlled by `allowAutomaticSingleRunInference` / `disallowAutomaticSingleRunInference` is not a 100% exact science (even though we really wish it could be). ESLint doesn't provide us the APIs we'd need: https://github.com/eslint/rfcs/pull/102. That's why we're phrasing it as _"inference"_.

When you call things through the manual API, you're taking on the burden of informing the parser of how things are meant to be. We can't reasonably infer how you're going to use the API. For example, you might add:

```js
await new Promise(resolve => setTimeout(resolve, 5000));
await eslint.lintFiles('test.vlue');
```

...after the end of the repro's `lint.js`.

So, _"working as intended"_, even though we're not too pleased about the _"intended"_. Thanks for filing though! This was helpful to reinforce our irritation at having to do these imprecise inferences in the first place. 🫠
